### PR TITLE
Extend Symbol with precise identifier span and two-provenance model

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- Adds precise identifier span support to Symbol via an optional name_span field.
- When present, name_span points at the exact identifier token; otherwise span remains coarse.
- AST nodes (Function, Relation, TypeDef) gain name_span() accessor; parser helpers added.
- Semantic model construction and traversal propagate name_span to new symbols where CST data is present.
- Lint diagnostics (e.g., unused-variable) prefer name_span when available.
- Documents and tests updated to describe and exercise the two-provenance model (span and name_span).

## Changes
- Core
  - Symbol: add name_span: Option<Span> and name_span() accessor

- AST
  - Function, Relation, TypeDef: implement name_span() using shared helper
  - Add find_identifier_span helper in parser/ast/mod.rs

- Parser
  - New helper find_identifier_span and related range search

- Semantic model builder/traversal
  - Wire name_span into SymbolSpec creation for Declarations
  - Collect and attach name_span for rule-head and binding symbols
  - Propagate name_span through collect_* where CST available

- Lint rules
  - unused_variable uses name_span for diagnostics when present

- Tests
  - New tests in src/sema/name_span.rs and integrated into sema/tests.rs

- Documentation
  - Update docs/ddlint-design.md to describe new two-provenance model (span and name_span)
- Docs: Additional parsing notes updated in docs/parser-implementation-notes.md

- Tests plan
  - [x] Unit tests for identifier spans
  - [x] Regression checks

### Notes
- name_span is optional to preserve backward compatibility; existing spans are preserved for diagnostics when precise spans are unavailable (e.g., parse-time SemanticRule values).
- The changes include an additive update to the semantic tests suite, with dedicated tests asserting the presence and correctness of name_span-derived spans.

Closes #253

📎 Task: https://www.devboxer.com/task/e372029c-5fe5-45f1-840c-7a52066a4fcf

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[src/parser/ast/mod.rs](https://github.com/leynos/ddlint/pull/256/files/fe9aff310ac4721689f5ac771f29c32b2e20faec#diff-1c67a302c8993b62cfec6d00e94c7c6a230f359c6ddea348f55ec261de51caa9)

Comment on file

~~~rust
pub(crate) fn find_identifier_span_in_range(
    syntax: &SyntaxNode<DdlogLanguage>,
    search_span: &Span,
    name: &str,
) -> Option<Span> {
    for child in syntax.children_with_tokens() {
        match child {
            NodeOrToken::Token(token)
                if token.kind() == SyntaxKind::T_IDENT && token.text() == name =>
            {
                let token_span = text_range_to_span(token.text_range());
                if span_contains(search_span, &token_span) {
                    return Some(token_span);
                }
            }
            NodeOrToken::Node(node) => {
                if !ranges_overlap(search_span, &text_range_to_span(node.text_range())) {
                    continue;
                }
                if let Some(span) = find_identifier_span_in_range(&node, search_span, name) {
                    return Some(span);
                }
            }
            NodeOrToken::Token(_) => {}
        }
    }

    None
}
~~~

❌ New issue: [Deep, Nested Complexity](https://codescene.io/projects/69278/delta?repo-id=1160463&review-id=256&biomarker=Deep%2C+Nested+Complexity&filename=src%2Fparser%2Fast%2Fmod.rs&method=find_identifier_span_in_range)
find_identifier_span_in_range has a nested complexity depth of 4, threshold = 4

File: src/sema/traverse.rs [hunk 1]
```diff
@@ -23,6 +26,18 @@ impl SemanticModelBuilder {
...
             for head in heads {
                 self.collect_head_expr(&head.atom, ctx);
+                for binding_name in collect_head_binding_names(&head.atom) {
+                    self.declare_symbol(SymbolSpec {
+                        name: binding_name.clone(),
+                        kind: DeclarationKind::RuleBinding,
+                        origin: ctx.origin,
+                        scope: ctx.scope,
+                        span: ctx.span.clone(),
+                        name_span: find_identifier_span(rule.syntax(), &binding_name),
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/sema/traverse.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[src/sema/traverse.rs](https://github.com/leynos/ddlint/pull/256/files/1ade13736a2b16997cefb3312df8d045da66b174#diff-070f23078955fd2db12ebb46d8c63c9613802cbdf4a0767e361d5138546c12ef)

Comment on file

~~~rust
    fn declare_pattern_binding(
        &mut self,
        binding_name: &str,
        origin: SymbolOrigin,
        scope: ScopeId,
        span: &Span,
        syntax: Option<&SyntaxNode<crate::DdlogLanguage>>,
        visible_from_rule_order: usize,
    ) {
        self.declare_symbol(SymbolSpec {
            name: binding_name.to_string(),
            kind: DeclarationKind::RuleBinding,
            origin,
            scope,
            span: span.clone(),
            name_span: syntax.and_then(|syntax| find_identifier_span(syntax, binding_name)),
            source_order: self.symbols.len(),
            visible_from_rule_order,
        });
    }
~~~

❌ New issue: [Excess Number of Function Arguments](https://codescene.io/projects/69278/delta?repo-id=1160463&review-id=256&biomarker=Excess+Number+of+Function+Arguments&filename=src%2Fsema%2Ftraverse.rs&method=SemanticModelBuilder.declare_pattern_binding)
SemanticModelBuilder.declare_pattern_binding has 7 arguments, max arguments = 4

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

[src/sema/tests/name_span.rs](https://github.com/leynos/ddlint/pull/256/files/63e8c597fa4d8ea42d187773d8c77baa49cf8bcb#diff-be64364f71bb8c9328cdb22ad1d76dee2d7a6e07812d64c3abba2b2b2de48acf)

Comment on file

~~~rust
fn rule_head_bindings_capture_identifier_name_spans() {
    let source = "Output(Output) :- Output(_).";
    let parsed = parse_ok(source);
    let semantic_model = super::super::build(&parsed);
    let symbol = find_symbol(
        &semantic_model,
        "Output",
        DeclarationKind::RuleBinding,
        SymbolOrigin::RuleHead,
    );

    let name_span = symbol
        .name_span()
        .unwrap_or_else(|| panic!("missing head binding name_span"));

    assert_eq!(span_text(source, name_span), "Output");
    assert_eq!(name_span.start, 7);
    assert_eq!(span_text(source, symbol.span()), source);
}
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/69278/delta?repo-id=1160463&review-id=256&biomarker=Code+Duplication&filename=src%2Fsema%2Ftests%2Fname_span.rs&method=)
The module contains 3 functions with similar structure: assignment_bindings_capture_identifier_name_spans,for_loop_bindings_capture_identifier_name_spans,rule_head_bindings_capture_identifier_name_spans

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

## Code Duplication
src/sema/tests/name_span.rs: 

### What lead to degradation?
The module contains 3 functions with similar structure: assignment_bindings_capture_identifier_name_spans,for_loop_bindings_capture_identifier_name_spans,rule_head_bindings_capture_identifier_name_spans

### Why does this problem occur?
Duplicated code often leads to code that's harder to change since the same logical change has to be done in multiple functions. More duplication gives lower code health.

### How to fix it?
A certain degree of duplicated code might be acceptable. The problems start when it is the same behavior that is duplicated across the functions in the module, ie. a violation of the Don't Repeat Yourself (DRY) principle. DRY violations lead to code that is changed together in predictable patterns, which is both expensive and risky. DRY violations can be identified using CodeScene's X-Ray analysis to detect clusters of change coupled functions with high code similarity. [Read More](https://codescene.com/blog/software-revolution-part3/)
Once you have identified the similarities across functions, look to extract and encapsulate the concept that varies into its own function(s). These shared abstractions can then be re-used, which minimizes the amount of duplication and simplifies change.

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

### ❌ Failed checks (3 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Docstring Coverage | ⚠️ Warning | Docstring coverage is 66.67% which is insufficient. The required threshold is 80.00%. | Write docstrings for the functions missing them to satisfy the coverage threshold. |
| Testing | ⚠️ Warning | Pull request introduces significant new functionality with incomplete test coverage for critical components including binding span collection logic and new AST accessor methods. | Add unit tests for collect\_head\_binding\_spans and should\_skip\_binding\_ident helper functions; test new name\_span() methods on Function, Relation, and TypeDef AST nodes; expand find\_identifier\_span tests for edge cases and nested structures. |
| Developer Documentation | ⚠️ Warning | PR introduces significant internal APIs and architectural changes without documenting implementation-level details in the designated implementation guide. | Update docs/parser-implementation-notes.md to document the find\_identifier\_span mechanism, RuleHead::binding\_spans computation, and syntax parameter threading through semantic builder functions. |
